In [149]:
## Current path display

import os
current_path = os.getcwd()
print(current_path)

d:\05_Projekte\04_Research\08_ManuBrain\01_Lab\Demo-Cell-Classification\02_Preprocessing


In [151]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import timedelta  

## Functions 

In [153]:
# return tag from data
def fc_tag (data,index):
    return data.at[index,'d'].split('=')[2] # [2] equivalent a tag column data en txt file

# update tag in main table
def fc_tag_update(data, tag):
    for i in range(len(tag)):
        i = i + 1 
        data = data.replace({'PrimaryKey':i}, fc_tag(tag,i), regex = False)
    return data

#Sync TimeStamp

def fc_sync_timestamp2(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for date in data[ServerTimeStamp]:
        source = datetime.fromisoformat(str(date).replace('Z','+00:00'))
        date_sync = source - delta_time
        data[ServerTimeStamp] = date_sync
    return data

def fc_sync_timestamp(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for i in range(len(data)): 
        source = datetime.fromisoformat(str(data.at[i,ServerTimeStamp]).replace('Z','+00:00'))
        date_sync = source - delta_time
        data.at[i,ServerTimeStamp] = date_sync
    return data

## Paths from data set

In [155]:
date = 201030
uc_num = 5

path = str(date) + '/use_case' + str(uc_num)

# Conveyor_belt
path_data_conv = '../01_Data/' + path + '/Conveyor Speed/DataLogger.csv'
path_tag_conv = '../01_Data/' + path + '/Conveyor Speed/tagconfig.txt'

# Gate
path_data_gate = '../01_Data/' + path + '/Gate/DataLogger.csv'
path_tag_gate = '../01_Data/' + path + '/Gate/tagconfig.txt'

# Quality_Control
path_data_qc = '../01_Data/' + path + '/Quality Control/DataLogger.csv'
path_tag_qc = '../01_Data/' + path + '/Quality Control/tagconfig.txt'

# Robot_Position
path_data_rp = '../01_Data/' + path + '/Robot Position/DataLogger.csv'
path_tag_rp = '../01_Data/' + path + '/Robot Position/tagconfig.txt'

# Store
path_data_st = '../01_Data/' + path + '/Store/DataLogger.csv'
path_tag_st = '../01_Data/' + path + '/Store/tagconfig.txt'

# Flag
path_data_flag = '../01_Data/' + path + '/Flag/DataLogger.csv'
path_tag_flag = '../01_Data/' + path + '/Flag/tagconfig.txt'

# Processed data

path_data_save = '../01_Data/' + path + '/Processed/'
path_sample_save = '../01_Data/' + path + '/Processed/Samples/'


## Import data from csv

In [157]:
# Conveyor_belt
data_conv =  pd.read_csv(path_data_conv, delimiter = ',')
#df_conv = pd.DataFrame(pd.read_csv(path_data_conv, delimiter = ','))

# Gate
data_gate =  pd.read_csv(path_data_gate, delimiter = ',')
#df_gate = pd.DataFrame(pd.read_csv(path_data_gate, delimiter = ','))

# Quality_Control
data_qc =  pd.read_csv(path_data_qc, delimiter = ',')
#df_qc = pd.DataFrame(pd.read_csv(path_data_qc, delimiter = ','))

# Robot_Position
data_rp =  pd.read_csv(path_data_rp, delimiter = ',')
#df_rp = pd.DataFrame(pd.read_csv(path_data_rp, delimiter = ','))

# Store
data_st =  pd.read_csv(path_data_st, delimiter = ',')
#df_st = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

# Flag
data_flag =  pd.read_csv(path_data_flag, delimiter = ',')
#df_flag = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

## Import data tags

In [159]:
# Conveyor_belt
tag_conv =  pd.read_fwf(path_tag_conv, sep=" ", header=None, names=['a','b','c','d'])
tag_conv = tag_conv.drop([0]) # delete first row (trash)

# Gate
tag_gate =  pd.read_fwf(path_tag_gate, sep=" ", header=None, names=['a','b','c','d'])
tag_gate = tag_gate.drop([0]) # delete first row (trash)

# Quality_Control
tag_qc =  pd.read_fwf(path_tag_qc, sep=" ", header=None, names=['a','b','c','d'])
tag_qc = tag_qc.drop([0]) # delete first row (trash)

# Robot_Position
tag_rp =  pd.read_fwf(path_tag_rp, sep=" ", header=None, names=['a','b','c','d'])
tag_rp = tag_rp.drop([0]) # delete first row (trash)

# Robot_Position
tag_st =  pd.read_fwf(path_tag_st, sep=" ", header=None, names=['a','b','c','d'])
tag_st = tag_st.drop([0]) # delete first row (trash)

# Flag
tag_flag =  pd.read_fwf(path_tag_flag, sep=" ", header=None, names=['a','b','c','d'])
tag_flag = tag_flag.drop([0]) # delete first row (trash)

## Update tags

In [161]:
# Conveyor_belt
data_conv = fc_tag_update(data_conv, tag_conv)

# Gate
data_gate = fc_tag_update(data_gate, tag_gate)

# Quality_Control
data_qc = fc_tag_update(data_qc, tag_qc)

# Robot_Position
data_rp = fc_tag_update(data_rp, tag_rp)

# Store
data_st = fc_tag_update(data_st, tag_st)

# Flag
data_flag = fc_tag_update(data_flag, tag_flag)

## Syncronize ServerTimeStamp

In [163]:
delta_time_PLC = timedelta(seconds=67)
delta_time_SIM = timedelta(seconds=0)

In [164]:
data_conv = fc_sync_timestamp(data_conv,'ServerTimeStamp',delta_time_PLC)
data_conv

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0         27   0x0004    300  0x002f0000  2020-10-30T12:46:18.868Z   
1         28   0x0004    363  0x002f0000  2020-10-30T12:46:18.868Z   
2         29   0x0004    380  0x002f0000  2020-10-30T12:46:18.868Z   

                    ServerTimeStamp  
0  2020-10-30 12:45:11.868000+00:00  
1  2020-10-30 12:45:11.868000+00:00  
2  2020-10-30 12:45:11.868000+00:00

In [165]:
data_gate = fc_sync_timestamp(data_gate,'ServerTimeStamp',delta_time_SIM)
data_gate

PrimaryKey DataType Value  StatusCode           SourceTimeStamp  \
0       Position   0x000b   0.3  0x00000000  2020-10-30T12:45:10.206Z   
1       Position   0x000b   0.3  0x00000000  2020-10-30T12:45:10.506Z   
2           Safe   0x0001  true  0x00000000  2020-10-30T12:45:10.206Z   
3           Safe   0x0001  true  0x00000000  2020-10-30T12:45:10.506Z   
4          Speed   0x000b     0  0x00000000  2020-10-30T12:45:10.206Z   
...          ...      ...   ...         ...                       ...   
18661      Speed   0x000b     0  0x00000000  2020-10-30T13:16:15.906Z   
18662      Speed   0x000b     0  0x00000000  2020-10-30T13:16:16.206Z   
18663   Position   0x000b   0.3  0x00000000  2020-10-30T13:16:16.506Z   
18664       Safe   0x0001  true  0x00000000  2020-10-30T13:16:16.506Z   
18665      Speed   0x000b     0  0x00000000  2020-10-30T13:16:16.506Z   

                        ServerTimeStamp  
0      2020-10-30 12:45:10.206000+00:00  
1      2020-10-30 12:45:10.506000+00:00  
2      2020-10-30 12:45:10.206000+00:00  
3      2020-10-30 12:45:10.506000+00:00  
4      2020-10-30 12:45:10.206000+00:00  
...                                 ...  
18661  2020-10-30 13:16:15.906000+00:00  
18662  2020-10-30 13:16:16.206000+00:00  
18663  2020-10-30 13:16:16.506000+00:00  
18664  2020-10-30 13:16:16.506000+00:00  
18665  2020-10-30 13:16:16.506000+00:00  

[18666 rows x 6 columns]

In [166]:
data_qc = fc_sync_timestamp(data_qc,'ServerTimeStamp',delta_time_SIM)
#data_qc

data_qc['Value'] = data_qc['Value'].replace(2,1)

# for row in data_qc.itertuples():
#     # current_row = data_qc[row_no, :]
#     if row.PrimaryKey == 'ResultCode':
#         if row.Value == 2:
#             row.Value = 1
#             print(row.Value)

data_qc

PrimaryKey DataType     Value  StatusCode           SourceTimeStamp  \
0      DistanceAbs   0x000b  0.777450  0x00000000  2020-10-30T12:45:14.303Z   
1      DistanceAbs   0x000b  0.777450  0x00000000  2020-10-30T12:45:14.603Z   
2        DistanceX   0x000b  0.433815  0x00000000  2020-10-30T12:45:14.303Z   
3        DistanceX   0x000b  0.433815  0x00000000  2020-10-30T12:45:14.603Z   
4        DistanceY   0x000b  0.645161  0x00000000  2020-10-30T12:45:14.303Z   
...            ...      ...       ...         ...                       ...   
24843   ResultCode   0x0006  0.000000  0x00000000  2020-10-30T13:16:17.303Z   
24844  DistanceAbs   0x000b  0.645161  0x00000000  2020-10-30T13:16:17.603Z   
24845    DistanceX   0x000b  0.000000  0x00000000  2020-10-30T13:16:17.603Z   
24846    DistanceY   0x000b  0.645161  0x00000000  2020-10-30T13:16:17.603Z   
24847   ResultCode   0x0006  0.000000  0x00000000  2020-10-30T13:16:17.603Z   

                        ServerTimeStamp  
0      2020-10-30 12:45:14.303000+00:00  
1      2020-10-30 12:45:14.603000+00:00  
2      2020-10-30 12:45:14.303000+00:00  
3      2020-10-30 12:45:14.603000+00:00  
4      2020-10-30 12:45:14.303000+00:00  
...                                 ...  
24843  2020-10-30 13:16:17.303000+00:00  
24844  2020-10-30 13:16:17.603000+00:00  
24845  2020-10-30 13:16:17.603000+00:00  
24846  2020-10-30 13:16:17.603000+00:00  
24847  2020-10-30 13:16:17.603000+00:00  

[24848 rows x 6 columns]

In [167]:
data_rp = fc_sync_timestamp(data_rp,'ServerTimeStamp',delta_time_PLC)
data_rp

PrimaryKey DataType       Value  StatusCode           SourceTimeStamp  \
0            527   0x0007  3198142330  0x00000000  2020-10-30T12:46:26.379Z   
1            527   0x0007  3198142877  0x00000000  2020-10-30T12:46:26.622Z   
2            528   0x0007  3187229088  0x00000000  2020-10-30T12:46:26.382Z   
3            528   0x0007  3187230597  0x00000000  2020-10-30T12:46:26.622Z   
4            529   0x0007  1053889550  0x00000000  2020-10-30T12:46:26.382Z   
...          ...      ...         ...         ...                       ...   
22054        527   0x0007  3198146987  0x00000000  2020-10-30T13:17:28.619Z   
22055        528   0x0007  3188627387  0x00000000  2020-10-30T13:17:28.369Z   
22056        528   0x0007  3188627451  0x00000000  2020-10-30T13:17:28.619Z   
22057        529   0x0007  1051184195  0x00000000  2020-10-30T13:17:28.366Z   
22058        529   0x0007  1051184101  0x00000000  2020-10-30T13:17:28.616Z   

                        ServerTimeStamp  
0      2020-10-30 12:45:19.379000+00:00  
1      2020-10-30 12:45:19.622000+00:00  
2      2020-10-30 12:45:19.382000+00:00  
3      2020-10-30 12:45:19.622000+00:00  
4      2020-10-30 12:45:19.382000+00:00  
...                                 ...  
22054  2020-10-30 13:16:21.619000+00:00  
22055  2020-10-30 13:16:21.369000+00:00  
22056  2020-10-30 13:16:21.619000+00:00  
22057  2020-10-30 13:16:21.366000+00:00  
22058  2020-10-30 13:16:21.616000+00:00  

[22059 rows x 6 columns]

In [168]:
data_st = fc_sync_timestamp(data_st,'ServerTimeStamp',delta_time_SIM)
data_st

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0      DiscSize   0x0003      0  0x00000000  2020-10-30T12:45:17.980Z   
1      DiscSize   0x0003      1  0x00000000  2020-10-30T12:45:18.280Z   
2      DiscSize   0x0003      0  0x00000000  2020-10-30T12:45:18.580Z   
3      DiscSize   0x0003      0  0x00000000  2020-10-30T12:45:18.880Z   
4      DiscSize   0x0003      0  0x00000000  2020-10-30T12:45:19.180Z   
...         ...      ...    ...         ...                       ...   
6203   DiscSize   0x0003      0  0x00000000  2020-10-30T13:16:18.880Z   
6204   DiscSize   0x0003      0  0x00000000  2020-10-30T13:16:19.180Z   
6205   DiscSize   0x0003      0  0x00000000  2020-10-30T13:16:19.480Z   
6206   DiscSize   0x0003      0  0x00000000  2020-10-30T13:16:19.780Z   
6207   DiscSize   0x0003      0  0x00000000  2020-10-30T13:16:20.080Z   

                       ServerTimeStamp  
0     2020-10-30 12:45:17.980000+00:00  
1     2020-10-30 12:45:18.280000+00:00  
2     2020-10-30 12:45:18.580000+00:00  
3     2020-10-30 12:45:18.880000+00:00  
4     2020-10-30 12:45:19.180000+00:00  
...                                ...  
6203  2020-10-30 13:16:18.880000+00:00  
6204  2020-10-30 13:16:19.180000+00:00  
6205  2020-10-30 13:16:19.480000+00:00  
6206  2020-10-30 13:16:19.780000+00:00  
6207  2020-10-30 13:16:20.080000+00:00  

[6208 rows x 6 columns]

In [169]:
data_flag = fc_sync_timestamp(data_flag,'ServerTimeStamp',delta_time_PLC)
data_flag

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          561   0x0001   true  0x00000000  2020-10-30T12:46:18.868Z   
1          560   0x0004      0  0x00000000  2020-10-30T12:46:18.870Z   
2          560   0x0004      1  0x00000000  2020-10-30T12:46:21.872Z   
3          561   0x0001  false  0x00000000  2020-10-30T12:46:22.368Z   
4          560   0x0004      2  0x00000000  2020-10-30T12:46:28.121Z   
..         ...      ...    ...         ...                       ...   
535        561   0x0001  false  0x00000000  2020-10-30T13:17:07.621Z   
536        560   0x0004      0  0x00000000  2020-10-30T13:17:13.870Z   
537        561   0x0001   true  0x00000000  2020-10-30T13:17:14.620Z   
538        560   0x0004      1  0x00000000  2020-10-30T13:17:21.372Z   
539        561   0x0001  false  0x00000000  2020-10-30T13:17:22.120Z   

                      ServerTimeStamp  
0    2020-10-30 12:45:11.868000+00:00  
1    2020-10-30 12:45:11.870000+00:00  
2    2020-10-30 12:45:14.872000+00:00  
3    2020-10-30 12:45:15.368000+00:00  
4    2020-10-30 12:45:21.121000+00:00  
..                                ...  
535  2020-10-30 13:16:00.621000+00:00  
536  2020-10-30 13:16:06.870000+00:00  
537  2020-10-30 13:16:07.620000+00:00  
538  2020-10-30 13:16:14.372000+00:00  
539  2020-10-30 13:16:15.120000+00:00  

[540 rows x 6 columns]

## Merge Data Sets

In [170]:
data_set = [data_conv, data_gate, data_qc, data_rp, data_st, data_flag]
data_merge = pd.concat(data_set, keys=['conv', 'gate', 'qc', 'rp', 'store', 'flag'])
# Sort the data_set by SourceTimeStamp, reset index and delete aux column
data_merge = data_merge.sort_values(by='ServerTimeStamp').reset_index().drop('level_1',1)
#del data_merge['level_1'] #delete aux column

In [171]:
data_merge

level_0 PrimaryKey DataType       Value  StatusCode  \
0        gate   Position   0x000b         0.3  0x00000000   
1        gate       Safe   0x0001        true  0x00000000   
2        gate      Speed   0x000b           0  0x00000000   
3        gate      Speed   0x000b           0  0x00000000   
4        gate   Position   0x000b         0.3  0x00000000   
...       ...        ...      ...         ...         ...   
72319      rp        527   0x0007  3198146937  0x00000000   
72320      rp        528   0x0007  3188627387  0x00000000   
72321      rp        529   0x0007  1051184101  0x00000000   
72322      rp        527   0x0007  3198146987  0x00000000   
72323      rp        528   0x0007  3188627451  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
1      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
2      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
3      2020-10-30T12:45:10.506Z  2020-10-30 12:45:10.506000+00:00  
4      2020-10-30T12:45:10.506Z  2020-10-30 12:45:10.506000+00:00  
...                         ...                               ...  
72319  2020-10-30T13:17:28.369Z  2020-10-30 13:16:21.369000+00:00  
72320  2020-10-30T13:17:28.369Z  2020-10-30 13:16:21.369000+00:00  
72321  2020-10-30T13:17:28.616Z  2020-10-30 13:16:21.616000+00:00  
72322  2020-10-30T13:17:28.619Z  2020-10-30 13:16:21.619000+00:00  
72323  2020-10-30T13:17:28.619Z  2020-10-30 13:16:21.619000+00:00  

[72324 rows x 7 columns]

## Export Data to csv

In [172]:
data_merge.to_csv(path_data_save + 'data_merge.csv')
print('The data is saved at ' + path_data_save + 'data_merge.csv')

The data is saved at ../01_Data/201030/use_case5/Processed/data_merge.csv


In [173]:
len(data_merge)

72324

## Determine the starting point of data set

#Based on change of DistanceAbs in Data- Set
var_actual = 0
run_list = [0]
for i in range(len(data_merge)):
    var_previous = var_actual
    if data_merge.at[i,'PrimaryKey'] == 'DistanceAbs':  # if 'PrimaryKey' value = 'DistanceAbs':
        var_actual = data_merge.at[i,'Value'] # value of current DistanceAbs
    if var_previous != var_actual:
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample    

# Based on grabbing the white disk
var_actual = 99
disk_actual = 99 # Initial value
run_list = [0]
disk = False
storage_down = False
rob_x = 0
rob_y = 0
rob_z = 0
rob_pos = 0
pos = [0]*(len(data_merge)-2)
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    pos[i] = 0
    if data_merge.at[i,'PrimaryKey'] == 'DiscSize':  # if 'PrimaryKey' value = 'DiskSize':
        disk_previous = disk_actual
        disk_actual = data_merge.at[i,'Value'] # value of current DiskSize
        #print(disk_previous, disk_actual, i)
        if disk_previous != 2 and disk_actual == 2: # Varialbe 0: unkown; 1: Black disk; 2: white disk
            disk = True
            #print('New white disk here ' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '527':  # if 'PrimaryKey' value = 'x coordinate':
        rob_x = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('x = ' + str(rob_x) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '528':  # if 'PrimaryKey' value = 'y coordinate':
        rob_y = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('y = ' + str(rob_y) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '529':  # if 'PrimaryKey' value = 'z coordinate':
        rob_z = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('z = ' + str(rob_z) + '--' + str(i))
    if rob_x > 3100000000 and rob_y > 3100000000 and rob_z < 1051200000: #Robot in storage. Picking up the disk
        rob_pos = 1
        pos[i] = 1
        #print('pos 1 - ' + str(i))
    if rob_x < 1052000000 and rob_y > 3100000000 and rob_z < 1052000000 and rob_z > 1051200000 : #Robot in storage. Placing down the disk
        rob_pos = 2
        pos[i] = 2
        #print('pos 2 - ' + str(i))
    #print(rob_pos)
    if rob_pos == 1: # Robot picking up a disk
        storage_down = True
        #print(storage_down,i)  
    if disk == True and storage_down == True:
        #var_actual = rob_pos
        disk = False
        storage_down = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
    #if var_previous != var_actual:
        #run_list.append(i) #i append to list # This list will tell at how many i contains one sample   
 

In [174]:
# Based on grabbing the white disk
run_list = [0]
rob_pos_actual = 99
rob_pos_old = 0
gripper_actual = 99
gripper_old = 0
gripper = False
rob_pos = False
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    if data_merge.at[i,'PrimaryKey'] == '561':  # if 'PrimaryKey' value = 'gripper open (0), close (1)':
        gripper_old = gripper_actual
        gripper_actual = data_merge.at[i,'Value'] # value of current 'x coordinate'
        if gripper_old != gripper_actual:
            gripper = True
    if data_merge.at[i,'PrimaryKey'] == '560':  # if 'PrimaryKey' value = 'position of robot based in certain conditions':
        # 0: Black disk in store
        # 1: Black disk in gate
        # 2: White disk in store
        # 3: White disk in gate
        rob_pos_old = rob_pos_actual
        rob_pos_actual = int(data_merge.at[i,'Value']) # value of current 'x coordinate'
        if rob_pos_old != rob_pos_actual and rob_pos_actual == 2:
            rob_pos = True
    if rob_pos == True and gripper == True:
        rob_pos = False
        gripper = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
 

In [175]:
print(run_list) # Starting points. 
print(len(run_list))

[0, 243, 1305, 2357, 3420, 4471, 5530, 6581, 7644, 8696, 10393, 11464, 12515, 13577, 14629, 15688, 16751, 17803, 18866, 19918, 20982, 22033, 23092, 24156, 25213, 26266, 27328, 28380, 29443, 30495, 31558, 32610, 33670, 34722, 35784, 36836, 37905, 38957, 40020, 41071, 42134, 43862, 44912, 45975, 47027, 48086, 49137, 50200, 51252, 52315, 53367, 54430, 55482, 56541, 57593, 58656, 59708, 60767, 61819, 62882, 63934, 64997, 66049, 67113, 68165, 69225, 70276, 71339]
68


In [176]:
# data_merge.iloc[list(range(0,3))]

In [177]:
#df_robpos = pd.DataFrame(pos,columns=['RobPos']) # Position of robot. 0: not important; 1: Robot in storage. Picking up the disk; 2:Robot in storage. Placing down the disk
#data_merge = pd.concat([data_merge, df_robpos], axis=1, sort=False)
data_merge

level_0 PrimaryKey DataType       Value  StatusCode  \
0        gate   Position   0x000b         0.3  0x00000000   
1        gate       Safe   0x0001        true  0x00000000   
2        gate      Speed   0x000b           0  0x00000000   
3        gate      Speed   0x000b           0  0x00000000   
4        gate   Position   0x000b         0.3  0x00000000   
...       ...        ...      ...         ...         ...   
72319      rp        527   0x0007  3198146937  0x00000000   
72320      rp        528   0x0007  3188627387  0x00000000   
72321      rp        529   0x0007  1051184101  0x00000000   
72322      rp        527   0x0007  3198146987  0x00000000   
72323      rp        528   0x0007  3188627451  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
1      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
2      2020-10-30T12:45:10.206Z  2020-10-30 12:45:10.206000+00:00  
3      2020-10-30T12:45:10.506Z  2020-10-30 12:45:10.506000+00:00  
4      2020-10-30T12:45:10.506Z  2020-10-30 12:45:10.506000+00:00  
...                         ...                               ...  
72319  2020-10-30T13:17:28.369Z  2020-10-30 13:16:21.369000+00:00  
72320  2020-10-30T13:17:28.369Z  2020-10-30 13:16:21.369000+00:00  
72321  2020-10-30T13:17:28.616Z  2020-10-30 13:16:21.616000+00:00  
72322  2020-10-30T13:17:28.619Z  2020-10-30 13:16:21.619000+00:00  
72323  2020-10-30T13:17:28.619Z  2020-10-30 13:16:21.619000+00:00  

[72324 rows x 7 columns]

The list contains the point i when the value is different. To consider the point where the value remains, consider i-1

In [178]:
d = {} # Dictionary that holds the sub-data-sets given by the starting point
for i in range(len(run_list)-1): # Ignores the last value (set of values)
    list_range = list(range(run_list[i],run_list[i+1]))
    d['df{0}'.format(i)] = data_merge.iloc[list_range]
    
print('Dictionary lenght is:' + str(len(d)))

Dictionary lenght is:67


## Assure conv_speed per sample 

In [179]:
conv_1 = 0
conv_2 = 0
conv_3 = 0
conv1 = False
conv2 = False
conv3 = False
for j in range(len(d)): # as many samples saved in dictionary
    df = d['df'+ str(j)]
    for i in range(len(df)): # as many row in the dataFrame
        #print(df.iat[i,1])
        if df.iat[i,1] == '27': # conv1
            conv_1 = df.iat[i,3]
            conv1 = True # the value shows at least once in the data set
        if df.iat[i,1] == '28': # conv2
            conv_2 = df.iat[i,3]
            conv2 = True
        if df.iat[i,1] == '29': # conv3
            conv_3 = df.iat[i,3]
            conv3 = True
    if conv1 == False: # The value did not occur in this sample. Copy value from previous samples into this one
        dic = {'PrimaryKey' : '27', 'Value': str(conv_1)}
        df.loc[len(df)+1,'PrimaryKey'] = '27'
        df.loc[len(df),'Value'] = str(conv_1)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv2 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_2)}
        df.loc[len(df)+1,'PrimaryKey'] = '28'
        df.loc[len(df),'Value'] = str(conv_2)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv3 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_3)}
        df.loc[len(df)+1,'PrimaryKey'] = '29'
        df.loc[len(df),'Value'] = str(conv_3)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
   
    conv1 = False # reset conv for the next round
    conv2 = False
    conv3 = False
print(conv_1)
print(conv_2)
print(conv_3)

300
363
380


## Export individual data-sample to .csv

In [180]:
for i in range(len(d)):
    if i > 0: # always ignore the first sample due to data-acquisition misalligment
        d['df' + str(i)].to_csv(path_sample_save + 'sample_' + str(i) + '.csv')
print('The data is saved at ' + path_sample_save + 'sample_x.csv')

The data is saved at ../01_Data/201030/use_case5/Processed/Samples/sample_x.csv


## Testing area

In [181]:
print("done")

done


In [182]:
# list_range = list(range(run_list[0],run_list[1]))
# print(list_range)

In [183]:
# data_merge.iloc[list_range]

In [184]:
# data_rp.loc[:, 'PrimaryKey'] = '529_z'
# data_rp

In [185]:
# %matplotlib notebook 
# data_gate.head(400).plot(x='SourceTimeStamp', y= 'Value')
# plt.xticks(rotation=45)

In [186]:
# %matplotlib notebook 
# data_rp.head(200).plot( y= 'Value')
# plt.xticks(rotation=45)